### ISSUE TITLE
Changes to cursor position without mouse event

### ISSUE THREAD

**Issue by anntzer** [matplotlib/matplotlib#7581]
This both affects the current tool system and the MEP22 toolmanager.

The cursor position only gets updated when the canvas detects a cursor motion, which leads to problems 1) immediately after switching fullscreen and 2) if alt-tab'ing to change window focus, moving the cursor and alt-tab'ing back to the original canvas.  With either case, the x-y position printed in the status bar is incorrect and it is possible for keyboard shortcuts (e.g.) "g" or "l" to end up affecting a "wrong" axes.

There is a length comment regarding this in the tkagg backend implementation (`_update_pointer_position`) but even that backend doesn't handle this properly.

Ultimately I think(?) the correct way to solve this is to do away with `_lastx/_lasty` and just refetch the cursor position whenever a LocationEvent must be emitted.

**Comment by tacaswell** [comment]
Do any of the 'enter_*' events get triggered this way?  We may need to add a 'gets focus' event?

**Comment by anntzer** [comment]
This would not help the case of switching to fullscreen.

**Comment by tacaswell** [comment]
For full screen we have a `resize_event`

**Comment by anntzer** [comment]
Good point.  I think hooking both resize_event and focus_event works better than refetching the cursor position before emitting any LocationEvents as the latter mechanism is probably(?) insufficient to update the status bar text.

**Comment by fariza** [comment]
We had a similar problem with the pointer when changing tools by shortcut https://github.com/matplotlib/matplotlib/pull/3474

**Comment by anntzer** [comment]
Another case is when an animation changes e.g. the x or y limits.  To cover this case, as well as the previously mentioned ones, it looks like we could hook draw_event (which should get called upon resize too?) and focus_event.

**Comment by anntzer** [comment]
Triggering a cursor update on a resize_event seems easy enough to do and should hopefully(?) be uncontroversial.  Adding a focus_event may be a bit more work...

**Comment by anntzer** [comment]
Yes.  I agree with your analysis above that fixing this probably needs having a gains_focus event.
